ddl/backfilling.go

```diff
@@ -282,46 +291,59 @@ func (w *backfillWorker) handleBackfillTask(d *ddlCtx, task *reorgBackfillTask,
 		zap.Int("scanCount", result.scanCount),
 		zap.String("nextHandle", tryDecodeToHandleString(result.nextKey)),
 		zap.String("takeTime", time.Since(startTime).String()))
+	if ResultCounterForTest != nil && result.err == nil {
+		ResultCounterForTest.Add(1)
+	}
 	return result
 }
 
 func (w *backfillWorker) run(d *ddlCtx, bf backfiller, job *model.Job) {
 	logutil.BgLogger().Info("[ddl] backfill worker start",
 		zap.Stringer("type", w.tp),
 		zap.Int("workerID", w.id))
-	defer func() {
-		w.resultCh <- &backfillResult{err: dbterror.ErrReorgPanic}
-	}()
-	defer util.Recover(metrics.LabelDDL, "backfillWorker.run", nil, false)
+	var curTaskID int
+	defer util.Recover(metrics.LabelDDL, "backfillWorker.run", func() {
+		w.resultCh <- &backfillResult{taskID: curTaskID, err: dbterror.ErrReorgPanic}
+	}, false)
 	for {
-		task, more := <-w.taskCh
-		if !more {
+		exit := false
+		select {
+		case task, more := <-w.taskCh:
+			if !more {
+				exit = true
+				break
+			}
+			curTaskID = task.id
+			d.setDDLLabelForTopSQL(job)
+
+			logutil.BgLogger().Debug("[ddl] backfill worker got task", zap.Int("workerID", w.id), zap.String("task", task.String()))
+			failpoint.Inject("mockBackfillRunErr", func() {
+				if w.id == 0 {
+					result := &backfillResult{taskID: task.id, addedCount: 0, nextKey: nil, err: errors.Errorf("mock backfill error")}
+					w.resultCh <- result
+					failpoint.Continue()
+				}
+			})
+
+			failpoint.Inject("mockHighLoadForAddIndex", func() {
+				sqlPrefixes := []string{"alter"}
+				topsql.MockHighCPULoad(job.Query, sqlPrefixes, 5)
+			})
+
+			failpoint.Inject("mockBackfillSlow", func() {
+				time.Sleep(100 * time.Millisecond)
+			})
+
+			// Change the batch size dynamically.
+			w.batchCnt = int(variable.GetDDLReorgBatchSize())
+			result := w.handleBackfillTask(d, task, bf)
+			w.resultCh <- result
```

Comment:
Could we put [line312-315](https://github.com/pingcap/tidb/pull/38646/files#diff-1bbc7833321782b3a22a842b6237974b6a60cb07b49b246b85552ab5e915ed6eR312-R315) to here, and replace `break` to `return`? If it needs, we can put line348 to line306. Then we can remove `exit` related code.